crates/control/src/calibration_controller.rs

```diff
@@ -0,0 +1,299 @@
+
+                if let Some(measurement) = values_top {
+                    self.current_measurements.push(measurement);
+                }
+                if let Some(measurement) = values_bottom {
+                    self.current_measurements.push(measurement);
+                }
+
+                if goto_next_position_top || goto_next_position_bottom {
+                    Some(self.try_transition_to_look_at(*context.cycle_time))
+                } else {
+                    None
+                }
+            }
+            CalibrationCommand::PROCESS => {
+                info!(
+                    "Switching to process!, found {} measurements",
+                    self.current_measurements.len()
+                );
+
+                let solved_result = solve(
+                    Corrections::default(),
+                    self.current_measurements.clone(),
+                    *context.field_dimensions,
+                );
+
+                info!("Calibration complete! Corrections: {solved_result:?}");
+                self.last_calibration_corrections = Some(solved_result);
+                Some(CalibrationCommand::FINISH)
+            }
+            CalibrationCommand::FINISH => None,
+        }
+    }
+
+    fn try_transition_to_look_at(&mut self, dispatch_time: CycleTime) -> CalibrationCommand {
+        let next_look_at = self.get_next_look_at();
+
+        match next_look_at {
+            Some((target, camera)) => CalibrationCommand::LOOKAT {
+                camera,
+                target,
+                dispatch_time,
+            },
+            None => {
+                info!("\tNothing else to LOOKAT, goto PROCESS");
+                CalibrationCommand::PROCESS
+            }
+        }
+    }
+
+    fn get_next_look_at(&mut self) -> Option<(Point2<Ground>, Option<CameraPosition>)> {
+        let current_index = self.current_look_at_index;
+        self.current_look_at_index += 1;
+        if current_index < self.look_at_list.len() {
+            Some(self.look_at_list[current_index])
+        } else {
+            None
+        }
+    }
+}
+
+fn collect_filtered_values(
+    measurement_perception_input: &PerceptionInput<Vec<&CalibrationCaptureResponse<Measurement>>>,
+    original_dispatch_time: &CycleTime,
+) -> (Option<Measurement>, bool) {
+    let result = measurement_perception_input
+        .persistent
+        .iter()
+        .flat_map(|(_cycle_timestamp, measurements)| measurements.iter())
+        .find(|&measurement| match measurement {
+            CalibrationCaptureResponse::CommandRecieved {
+                dispatch_time,
+                output: _,
+            } => original_dispatch_time.start_time == dispatch_time.start_time,
+            CalibrationCaptureResponse::Idling => false,
+            CalibrationCaptureResponse::RetriesExceeded { dispatch_time } => {
+                original_dispatch_time.start_time == dispatch_time.start_time
+            }
+        });
```

Comment:
Yes, I think this would be fine as there can only be one response from the vision cycler.

That said, I should reconsider using `filter_map` as this filtering needs a lot of chaining.